### PR TITLE
util/goroutine_pool: add a goroutine pool package utilities

### PR DESCRIPTION
**Pull request by tiancaiamao**
Reuse goroutine in a pool can avoid `runtime.morestack`. It may be useful sometimes.

```
BenchmarkGoPool-4          	 5000000	       392 ns/op
BenchmarkGo-4              	10000000	       161 ns/op
BenchmarkMorestackPool-4   	 2000000	       706 ns/op
BenchmarkMoreStack-4       	  500000	      3018 ns/op
```

### CODE REVIEW

File: util/goroutine_pool/gp_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (shenli) [inline util/goroutine_pool/gp_test.go]: 2016 -> 2017

**Comment by tiancaiamao** [comment]
I've test in another branch, using pool mechanism can totally eliminate `runtime.morestack`.
PTAL @shenli @ngaut @winkyao

File: util/goroutine_pool/gp.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	p.status = statusIdle
+	pool.Unlock()
+}
+
+func (pool *Pool) alloc() *goroutine {
+	g := &goroutine{
+		ch:   make(chan func()),
+		pool: pool,
+	}
+	go func(g *goroutine) {
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: I think defining a method for `goroutine` is more clear.

File: util/goroutine_pool/gp.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	if ret == pool.tail {
+		pool.tail = head
+	}
+	pool.count--
+	pool.Unlock()
+	ret.next = nil
+	return ret
+}
+
+func (pool *Pool) put(p *goroutine) {
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: This is called by `goroutine` in background, I think make this a `goroutine` method is more reasonable.

File: util/goroutine_pool/gp.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	p.next = nil
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: Isn't it always nil, encured by `get`?

File: util/goroutine_pool/gp.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	g := &goroutine{
+		ch: make(chan func()),
+	}
+	go g.workLoop(pool)
+	return g
+}
+
+func (g *goroutine) put(pool *Pool) {
+	g.status = statusIdle
+	g.next = nil
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: Is it possible that `g.next` is not nil here?

File: util/goroutine_pool/gp.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	pool.Lock()
+	head := &pool.head
+	if head.next == nil {
+		pool.Unlock()
+		return pool.alloc()
+	}
+
+	ret := head.next
+	head.next = ret.next
+	if ret == pool.tail {
```
Comment (coocood) [inline util/goroutine_pool/gp.go]:
```
pool.count--
if pool.count == 0 {
  pool.tail = head
}
```
 is easier to read.

File: util/goroutine_pool/gp.go (hunk 5 above)
Comment (tiancaiamao) [reply util/goroutine_pool/gp.go]: I'm not sure.

File: util/goroutine_pool/gp.go (hunk 6 above)
Comment (tiancaiamao) [reply util/goroutine_pool/gp.go]: I didn't intent to maintain count, it's added later just for testing purpose.

File: util/goroutine_pool/gp.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+func (pool *Pool) Go(f func()) {
+	var g *goroutine
+	for {
+		g = pool.get()
+		if atomic.CompareAndSwapInt32(&g.status, statusIdle, statusInUse) {
+			break
+		}
+		// Status already changed from statusIdle => statusDying, delete this goroutine.
+		if atomic.LoadInt32(&g.status) == statusDying {
+			g.status = statusDead
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: It seems no body cares about this value any more.
It can be removed.

File: util/goroutine_pool/gp.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// Go works like go func(), but goroutines are pooled for reusing.
+// This strategy can avoid runtime.morestack, because pooled goroutine is already enlarged.
+func (pool *Pool) Go(f func()) {
+	var g *goroutine
+	for {
+		g = pool.get()
+		if atomic.CompareAndSwapInt32(&g.status, statusIdle, statusInUse) {
+			break
```
Comment (coocood) [inline util/goroutine_pool/gp.go]: I think just call
`g.ch <-f` and return here is easier to read.

Comment (tiancaiamao) [reply util/goroutine_pool/gp.go]: It's easier to read indeed, but there is a potential race condition.

File: util/goroutine_pool/gp_test.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+	gp.Lock()
+	count := gp.count
+	gp.Unlock()
+	if count != 1 {
+		t.Error("all goroutines should be recycled", count)
+	}
+}
+
+func TestRace(t *testing.T) {
+	gp := New(200 * time.Millisecond)
```
Comment (coocood) [inline util/goroutine_pool/gp_test.go]: Isn't the timeout too long to test race?

**Comment by coocood** [comment]
LGTM

**Pull request merged**